marimo/_plugins/ui/_impl/dataframes/handlers.py

```diff
@@ -0,0 +1,261 @@
+        else:
+            _assert_never(transform.aggregation)
+
+    @staticmethod
+    def handle_aggregate(
+        df: "pd.DataFrame", transform: AggregateTransform
+    ) -> "pd.DataFrame":
+        dict_of_aggs = {
+            column_id: transform.aggregations
+            for column_id in transform.column_ids
+        }
+        return df.agg(dict_of_aggs)  # type: ignore[arg-type]
+
+
+def apply_transforms(
+    df_prev: "pd.DataFrame", transforms: Transformations
+) -> "pd.DataFrame":
+    if transforms.transforms is None or len(transforms.transforms) == 0:
+        return df_prev
+    # make a new reference
+    df = df_prev
+    for transform in transforms.transforms:
+        df = TransformHandlers.handle(df, transform)
+    return df
+
+
+def _assert_never(value: NoReturn) -> NoReturn:
+    raise AssertionError(f"Unhandled value: {value} ({type(value).__name__})")
+
+
+def _coerce_value(dtype: Any, value: Any) -> Any:
+    import numpy as np
+
+    return np.array([value]).astype(dtype)[0]
+
+
+class TransformsContainer:
+    """
+    Keeps internal state of the last transformation applied to the dataframe.
+    So that we can incrementally apply transformations.
+    """
+
+    def __init__(self, df: "pd.DataFrame") -> None:
+        self._original_df = df
+        # The dataframe for the given transform.
+        self._snapshot_df = df
+        self._transforms: List[Transform] = []
+
+    def apply(
+        self, df: "pd.DataFrame", transform: Transformations
+    ) -> "pd.DataFrame":
+        """
+        Applies the given transformations to the dataframe.
+        """
+        # If differs from the original dataframe, then we need to start from
+        # the original dataframe.
+        if df is not self._original_df:
+            self._original_df = df
+            self._snapshot_df = apply_transforms(df, transform)
+            self._transforms = transform.transforms
+            return self._snapshot_df
+
+        # If the new transformations are a superset of the existing ones,
+        # then we can just apply the new ones to the snapshot dataframe.
+        if self._is_superset(transform):
+            transforms_to_apply = self._get_next_transformations(transform)
+            self._snapshot_df = apply_transforms(
+                self._snapshot_df, transforms_to_apply
+            )
+            self._transforms = transform.transforms
+            return self._snapshot_df
+
+        # If the new transformations are not a superset of the existing ones,
+        # then we need to start from the original dataframe.
+        else:
+            self._snapshot_df = apply_transforms(self._original_df, transform)
+            self._transforms = transform.transforms
+            return self._snapshot_df
+
+    def _is_superset(self, transforms: Transformations) -> bool:
+        """
+        Checks if the new transformations are a superset of the existing ones.
+        """
+        if len(self._transforms) == 0:
```

Comment:
nit: Python convention is to just write

```python
if not self._transforms:
   ...
```

Instead of explicitly checking length.

https://peps.python.org/pep-0008/

![image](https://github.com/marimo-team/marimo/assets/1994308/8bf080eb-3b7a-442f-a2a6-9dece7044ccc)

Reply:
can we enforce this with linting?

Reply:
There might be a rule in ruff, but I couldn't find it.